### PR TITLE
feat: add Nx.stack/2

### PR DESCRIPTION
**Pull request by polvalente**
This PR adds Nx.stack as suggested in #379

### CODE REVIEW

File: nx/lib/nx.ex [hunk 1]
```diff
@@ -7332,6 +7332,36 @@ defmodule Nx do
...
+        [1, 2, 3]
+      >
+
+      iex> Nx.stack([Nx.tensor(1), Nx.tensor(2)], name: :x)
+      #Nx.Tensor<
+        s64[x: 2]
+        [1, 2]
+      >
+  """
+  def stack(enumerable, opts \\ []) do
```
Comment (polvalente) [inline nx/lib/nx.ex:7356]: Should I validate that is is an enumerable somehow?

Also, should this be delegated to individual backends? It seems to me that this belongs here, but I might have missed something

Comment (seanmor5) [reply nx/lib/nx.ex:7356]: I would validate it's a list here, I think that makes sense

**Review by seanmor5 — APPROVED** [review]
Looks great!

File: nx/lib/nx.ex (hunk 1 above)
Comment (polvalente) [reply nx/lib/nx.ex:7356]: Technically this also works for MapSets or any other Enumerable which yields tensors when `Enum.map`ed

**Comment by polvalente** [comment]
@seanmor5 one last experiment I did:

```elixir
iex(7)> Enum.map(1..3, & &1 |> Nx.tensor() |> Nx.reshape({1, 1, 1})) |> Nx.stack()
#Nx.Tensor<
  s64[3][1][1][1]
  [
    [
      [
        [1]
      ]
    ],
    [
      [
        [2]
      ]
    ],
    [
      [
        [3]
      ]
    ]
  ]
>
iex(8)> Enum.map(1..3, & &1 |> Nx.tensor() |> Nx.reshape({1, 1, 1})) |> Nx.concatenate
#Nx.Tensor<
  s64[3][1][1]
  [
    [
      [1]
    ],
    [
      [2]
    ],
    [
      [3]
    ]
  ]
>
```

Is this a desired behavior, or should we only accept `{}`-shaped tensors for `Nx.stack`?

File: nx/lib/nx.ex (hunk 1 above)
Comment (josevalim) [reply nx/lib/nx.ex:7356]: Indeed but I think those use cases won't be common, so I would go with something like:

```suggestion
  def stack(tensors, opts \\ []) when is_list(tensors) do
```

**Review by josevalim — APPROVED** [review]
Just one small change and ship it!

**Pull request merged**